### PR TITLE
Refactor compiler

### PR DESCRIPTION
**Pull request by oritstone**

### CODE REVIEW

**Comment by oritstone** [comment]
@orius123 please review & merge

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompiler.java [hunk 1]
```diff
@@ -8,22 +8,28 @@
...
 
 import java.io.Serializable;
 import java.util.Map;
 import java.util.Set;
 
-//todo: Eliya - add JavaDoc
 public interface SlangCompiler {
 
+    /**
+     * Compile a Slang source & its path to a CompilationArtifact object
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompiler.java]: I would switch the & with the word 'and'.
make classes link with the `@link` javadocs property

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompiler.java [hunk 2]
```diff
@@ -8,22 +8,28 @@
...
+     * @param source the slang source file
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompiler.java]: not necessarily a file

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java [hunk 3]
```diff
@@ -41,53 +31,28 @@
...
-        ExecutionPlan executionPlan = compileToExecutionPlan(executable);
-
-        //and also create execution plans for all other dependencies
-        Map<String, ExecutionPlan> dependencies = convertMap(filteredDependencies, new Converter<Executable, ExecutionPlan>() {
-            @Override
-            public ExecutionPlan convert(Executable compiledExecutable) {
-                return compileToExecutionPlan(compiledExecutable);
+        //we transform also all of the files in the given path to model objects
+        Set<Executable> pathExecutables = new HashSet<>();
+        if(path != null && !path.isEmpty()) {
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java]: would use `CollectionUtils.isEmpty` for simplicity

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java [hunk 4]
```diff
@@ -41,53 +31,28 @@
...
+            for (SlangSource pathSource : path) {
+                pathExecutables.add(preCompile(pathSource));
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java]: would switch the for with `Lambda.map()` for simplicity

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/configuration/SlangCompilerSpringConfig.java [hunk 5]
```diff
@@ -30,4 +34,9 @@ public Yaml yaml(){
         return yaml;
     }
 
+    @Bean
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/configuration/SlangCompilerSpringConfig.java]: it is annotated with `@component`, why doesn't `packageScan` load it?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java [hunk 6]
```diff
@@ -220,11 +78,7 @@ public Executable preCompile(SlangSource source) {
...
         ParsedSlang parsedSlang = yamlParser.parse(source);
 
-        try {
-            //then we transform those maps to model objects
-            return transformToExecutable(parsedSlang);
-        } catch (Throwable ex){
-            throw new RuntimeException("Error compiling source: " + source.getName() + ". " + ex.getMessage(), ex);
-        }
+        // Then we transform the parsed Slang source to a Slang model
+        return slangModeller.createModel(parsedSlang);
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/SlangCompilerImpl.java]: who handles null for system properties?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/SlangModeller.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public interface SlangModeller {
+
+    /**
+     * Pre-compile a Slang file into an Executable
+     * @param parsedSlang the parsed Slang source file
+     * @return an Executable object, containing either a flow or an operations in the file
+     *          Returns null if the source contains Slang system properties
+     */
+    public Executable createModel(ParsedSlang parsedSlang);
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/SlangModeller.java]: It's an interface no need for `public` modifier

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ResultsTransformer.java [hunk 8]
```diff
@@ -14,6 +14,7 @@
  * Created by orius123 on 05/11/14.
  */
 
+import org.openscore.lang.compiler.modeller.transformers.Transformer;
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ResultsTransformer.java]: ?
isn't it in the same package?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/WorkFlowTransformer.java [hunk 9]
```diff
@@ -14,6 +14,7 @@
  * Created by orius123 on 05/11/14.
  */
 
+import org.openscore.lang.compiler.modeller.transformers.Transformer;
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/WorkFlowTransformer.java]: again, do you have to import it?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompiler.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.Set;
+
+/**
+ * Created by stoneo on 2/2/2015.
+ */
+public interface ExecutionPlanCompiler {
+
+
+    CompilationArtifact compile(Executable executable, Set<Executable> dependencies);
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompiler.java]: docs

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompilerImpl.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.Set;
+
+/**
+ * Created by stoneo on 2/2/2015.
+ */
+public class ExecutionPlanCompilerImpl implements ExecutionPlanCompiler{
+
+    public ExecutionPlanCompilerImpl() {
+        super();
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompilerImpl.java]: ?????

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompilerImpl.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    public CompilationArtifact compile(Executable executable, Set<Executable> dependencies) {
+        return null;
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ExecutionPlanCompilerImpl.java]: ???
what is this class??

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompiler.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public interface ScoreCompiler {
+
+    /**
+     * Compile an Executable & its path to a CompilationArtifact object
+     * @param source the executable source
+     * @param path a set of executables containing the source dependencies
+     * @return the compiled CompilationArtifact
+     */
+    public CompilationArtifact compile(Executable source, Set<Executable> path);
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompiler.java]: no need for `public`
again about the `&`

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Autowired
+    private DependenciesHelper dependenciesHelper;
+
+    @Override
+    public CompilationArtifact compile(Executable executable, Set<Executable> path) {
+
+        Map<String, Executable> filteredDependencies = new HashMap<>();
+        //we handle dependencies only if the file has imports
+        boolean hasDependencies = MapUtils.isNotEmpty(executable.getDependencies())
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: what about an operation with system properties?

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+     * Utility method that cast a {@link org.openscore.lang.compiler.modeller.model.Executable} to its subtype
+     * and create an {@link org.openscore.api.ExecutionPlan} for it
+     *
+     * @param executable the executable to create an {@link org.openscore.api.ExecutionPlan} for
+     * @return {@link org.openscore.api.ExecutionPlan} of the given {@link org.openscore.lang.compiler.modeller.model.Executable}
+     */
+    private ExecutionPlan compileToExecutionPlan(Executable executable) {
+        ExecutionPlan executionPlan;
+
+        if (executable.getType().equals(SlangTextualKeys.OPERATION_TYPE)) {
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: can we 'switch case' it?

**Comment by orius123** [comment]
in general except for ExecutionPlanCompiler class and interface, fell free to merge.
all the other comments are minor.
awesome work!! :astonished:

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java (hunk 14 above)
Comment (oritstone) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: System properties are currently not considered as dependencies at all.
We need to rethink it once we decide how to treat system props in general

**Pull request merged**